src/lib/Simplify.hs

```diff
@@ -525,21 +527,28 @@ simplifyOp op = case op of
   ScalarBinOp (ICmp Less ) x y -> ilt x y
   ScalarBinOp (ICmp Equal) x y -> ieq x y
   Select c x y -> select c x y
-  ProjMethod dict methodName -> do
-    Con (LabelCon methodName') <- return methodName
-    lookupSourceMap methodName' >>= \case
-      Just (UMethodVar  v') -> lookupEnv v' >>= \case
-        MethodBinding _ i _ -> do
-          return $ getProjection [i,1,0] dict
-      _ -> throw TypeErr "Not a method name"
-  ExplicitDict dictTy method -> do
-    TypeCon sourceName name params <- return dictTy
-    return $ DataCon sourceName name params 0 [PairVal (ProdVal []) (ProdVal [method])]
+  ProjMethod dict i -> projectDictMethod dict i
   _ -> emitOp op
 
 pattern IdentityReconAbs :: Abs binder ReconstructAtom n
 pattern IdentityReconAbs <- Abs _ IdentityRecon
 
+projectDictMethod :: Emits o => Atom o -> Int -> SimplifyM i o (Atom o)
+projectDictMethod d i = do
+  cheapNormalize d >>= \case
+    DictCon (InstanceDict instanceName args) -> dropSubst do
+      args' <- mapM simplifyAtom args
+      InstanceDef _ bs _ body <- lookupInstanceDef instanceName
+      let InstanceBody _ methods = body
+      let method = methods !! i
+      extendSubst (bs@@>(SubstVal <$> args')) $
+        simplifyBlock method
```

Comment:
Can you please explain how is it that we end up with `InstanceDict`s in the end? Why couldn't it be a superclass projection?

Reply:
If it were a superclass projection, then either you're projecting out of an instance, in which case you can immediately resolve that projection, or else you're projecting out of a var, but that's impossible because vars can't have `DictTy` types post-simplification.